snapcraft/models/project.py

```diff
@@ -1221,16 +1223,35 @@ def get_build_plan(self) -> list[BuildInfo]:
                     Platform.from_architectures(self.architectures)
                 )
 
-        for platform_entry, platform in self.platforms.items():
-            for build_for in platform.build_for or [SnapArch(platform_entry).value]:
-                for build_on in platform.build_on or [SnapArch(platform_entry).value]:
-                    build_infos.append(
-                        BuildInfo(
-                            platform=platform_entry,
-                            build_on=build_on,
-                            build_for=build_for,
-                            base=base,
-                        )
-                    )
+        platforms = cast(
+            Platforms,
+            {name: platform.marshal() for name, platform in self.platforms.items()},
+        )
 
-        return build_infos
+        # In _validate_mandatory_base, we ensure that the possible values of
+        # 'base' and 'snap_type' are narrowed so they'll always match one of
+        # the two overloads of get_platforms_snap_build_plan.  But, pyright and
+        # mypy aren't smart enough to realize this, so we need the type checker
+        # ignores.
+        # Also note that you have to put mypy's "type" ignore before the
+        # pyright ignore when they're on the same line, because only pyright is
+        # smart enough to register the ignore when it isn't the first comment
+        # on the line.
+        _validate_mandatory_base(self.base, self.type)
+        return [
+            BuildInfo(
+                platform=buildinfo.platform,
+                build_on=str(buildinfo.build_on),
+                build_for=str(buildinfo.build_for),
+                base=bases.BaseName(
+                    name=buildinfo.build_base.distribution,
+                    version=buildinfo.build_base.series,
+                ),
+            )
+            for buildinfo in snap.get_platforms_snap_build_plan(  # pyright: ignore[reportCallIssue]
+                base=self.base,  # type: ignore[arg-type]  # pyright: ignore[reportArgumentType]
```

Comment:
I really thought that:

```suggestion
                base=self.base,  # type: ignore[arg-type]
```

would silence pyright's reportArgumentType too?

Reply:
Hey it seems like you're right!